pkg/operators/ebpf/stats.go

```diff
@@ -0,0 +1,527 @@
+	if err != nil {
+		return progStat{}, err
+	}
+	defer prog.Close()
+
+	pi, err := prog.Info()
+	if err != nil {
+		return progStat{}, err
+	}
+
+	runtime, _ := pi.Runtime()
+	runcount, _ := pi.RunCount()
+
+	return progStat{
+		runtime:  uint64(runtime),
+		runcount: runcount,
+	}, nil
+}
+
+func (i *ebpfOperatorDataInstance) getStats() ([]stat, error) {
+	stats := make([]stat, 0)
+
+	mapSizes, err := bpfstats.GetMapsMemUsage()
+	if err != nil {
+		return nil, fmt.Errorf("getting map memory usage: %w", err)
+	}
+
+	// TODO: reduce lock contention
+	i.bpfOperator.mu.Lock()
+	defer i.bpfOperator.mu.Unlock()
+
+	oldStats := i.oldProgStats
+	i.oldProgStats = make(map[ebpf.ProgramID]progStat)
+
+	// emit consolidated stats for gadgets
+	if !i.programStats {
+		cache := make(map[ebpf.ProgramID]progStat)
+
+		for ctx, gadgetObjs := range i.bpfOperator.gadgetObjs {
+			stat := stat{
+				gadgetID:    ctx.ID(),
+				gadgetName:  ctx.Name(),
+				gadgetImage: ctx.ImageName(),
+			}
+
+			for _, id := range gadgetObjs.programIDs {
+				progStat, err := getProgramStats(cache, id)
+				if err != nil {
+					return nil, fmt.Errorf("getting program stats: %w", err)
+				}
+
+				i.oldProgStats[id] = progStat
+				oldProgStats := oldStats[id]
+
+				stat.runtime += progStat.runtime - oldProgStats.runtime
+				stat.runcount += progStat.runcount - oldProgStats.runcount
+			}
+
+			for _, mapID := range gadgetObjs.mapIDs {
+				stat.mapMemory += mapSizes[mapID]
+				stat.mapCount += 1
+			}
+
+			stats = append(stats, stat)
+		}
+
+		return stats, nil
+	}
+
+	// emit per ebpf programs stats
+	programToGadget := make(map[ebpf.ProgramID]operators.GadgetContext)
+	for ctx, gadgetObjs := range i.bpfOperator.gadgetObjs {
+		for _, id := range gadgetObjs.programIDs {
+			programToGadget[id] = ctx
+		}
+	}
+
+	var curID ebpf.ProgramID
+	var nextID ebpf.ProgramID
+	for {
+		nextID, err = ebpf.ProgramGetNextID(curID)
+		if err != nil {
+			if errors.Is(err, os.ErrNotExist) {
+				break
+			}
+			return nil, fmt.Errorf("getting next program ID: %w", err)
+		}
+		if nextID <= curID {
+			break
+		}
+		curID = nextID
+		prog, err := ebpf.NewProgramFromID(curID)
+		if err != nil {
+			continue
+		}
+		pi, err := prog.Info()
+		if err != nil {
+			prog.Close()
+			continue
+		}
+
+		stat := stat{}
+
+		id, _ := pi.ID()
+		stat.programID = uint32(id)
+		stat.programName = pi.Name
+		runtime, _ := pi.Runtime()
+		stat.runtime = uint64(runtime)
+		stat.runcount, _ = pi.RunCount()
+
+		i.oldProgStats[curID] = progStat{
+			runtime:  stat.runtime,
+			runcount: stat.runcount,
+		}
+
+		oldProgStat := oldStats[curID]
+		stat.runtime -= oldProgStat.runtime
+		stat.runcount -= oldProgStat.runcount
+
+		mapIDs, _ := pi.MapIDs()
+		for _, mapID := range mapIDs {
+			stat.mapMemory += mapSizes[mapID]
+			stat.mapCount += 1
+		}
+
+		// enrich with gadget information if they're part of a gadget
+		if ctx, ok := programToGadget[ebpf.ProgramID(id)]; ok {
```

Comment:
This only works if the gadget is executed by the same ig process as the one running the stats.

To make it work when running several ig processes, we have to use bpf_metadata:
https://qmonnet.github.io/whirl-offload/2021/09/23/bpftool-features-thread/#program-metadata

Could you file an issue for a future PR?

Reply:
TIL about bpf_metadat, I'll open the issue when merging this PR. Thanks!

Reply:
Ref #4317